### PR TITLE
rootfs: Add script for Debian base OS

### PR DESCRIPTION
**Pull request by mcgrof**
Add scripts to build a rootfs based on Debian.

Signed-off-by: Luis Chamberlain <mcgrof@kernel.org>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @mcgrof - thanks for raising!

We have a small issue though - @ydjainopensource has raised #144 to support Ubuntu and I think that since Ubuntu+Debian are so similar, it makes sense to land support for both as they should be able to share 99.99% of the config (see https://github.com/kata-containers/osbuilder/pull/144#discussion_r206095779). @ydjainopensource - wdyt?

On a general note, you'll see that the CI's have failed. If you look at the Travis log, you'll see that it's failing since our [`checkcommits`](https://github.com/kata-containers/tests/tree/master/cmd/checkcommits) tool is complaining there is no "`Fixes #XXX`" comment in you commit, and pointing you to:

- https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format

```
Found 1 commit between commit 5237f1e27b438e6bacb9b026c6d576408f6aaa4b and branch master
ERROR: No "Fixes" found
ERROR: checkcommits failed. See the document below for help on formatting
commits for the project.

https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format
```

To resolve this issue, you'd need to tweak your commit to reference a github issue using the magic `Fixes #XXX` comment. Something like the following, rebase and then re-push:

```
rootfs: Add script for Debian base OS

Add scripts to build a rootfs based on Debian.

Fixes #XXX.

Signed-off-by: Luis Chamberlain <mcgrof@kernel.org>
```

However, we need to resolve the Ubuntu vs. Debian issue first I think :smile:

**Comment by ydjainopensource** [comment]
@jodh-intel yes, the ubuntu and debian are almost similar. Just having ubuntu source debian or vice versa is much more simpler than maintaining two different scripts.

@mcgrof Your PR in its current state needs a lot of changes before it can be merged. I can't get it to run with USE_DOCKER=true. Can you make changes to your PR such that it would source the scripts for ubuntu. Then once #144 lands we can merge this one too.

Let me know if you wish to include any changes in #144.

File: rootfs-builder/debian/Dockerfile.in [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# Copyright (c) 2018 Luis Chamberlain <mcgrof@kernel.org>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+From debian:@OS_VERSION@
+
+@SET_PROXY@
+
+RUN apt-get update && apt-get install -y git systemd pkg-config gcc coreutils
+
```
Comment (ydjainopensource) [inline rootfs-builder/debian/Dockerfile.in]: You need to include curl and debootstrap here.

File: rootfs-builder/debian/config.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#
+# Copyright (c) 2018 Luis Chamberlain <mcgrof@kernel.org>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+OS_NAME="Debian"
+
+OS_VERSION=${OS_VERSION:-testing}
```
Comment (ydjainopensource) [inline rootfs-builder/debian/config.sh]: Ideally this should use the stable release

File: rootfs-builder/debian/rootfs_lib.sh [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	DEBOOTSTRAP_MIRROR="$MIRROR_LIST"
+	if [ "$LOCAL_FILE_MIRROR" != "" ]; then
+		DEBOOTSTRAP_MIRROR="$LOCAL_FILE_MIRROR"
+	fi
+
+	/usr/sbin/debootstrap \
+		$INCLUDE_EXTRA \
+		${OS_VERSION} \
+		$ROOTFS_DIR \
+		$DEBOOTSTRAP_MIRROR
```
Comment (ydjainopensource) [inline rootfs-builder/debian/rootfs_lib.sh]: You need to include `--arch` here. To allow cross platform builds.

File: README.md [hunk 4]
```diff
@@ -120,7 +120,7 @@ For further details, see [the tests documentation](tests/README.md).
 
 ## Platform-Distro Compatibility Matrix
 
-| | Alpine | CentOS | ClearLinux | EulerOS | Fedora | 
+| | Alpine | CentOS | ClearLinux | EulerOS | Fedora | Debian |
```
Comment (ydjainopensource) [inline README.md]: The table is not displayed properly on my pc. Can you confirm?

File: rootfs-builder/debian/rootfs_lib.sh [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	for i in ${EXTRA_PKGS} ; do
+		if [ "$ADD_PKGS" = "" ]; then
+			ADD_PKGS="$i"
+		else
+			ADD_PKGS="$ADD_PKGS,$i"
+		fi
+	done
+	INCLUDE_EXTRA=""
+	if [ "$ADD_PKGS" = "" ]; then
+		INCLUDE_EXTRA="--include=$ADD_PKGS"
```
Comment (ydjainopensource) [inline rootfs-builder/debian/rootfs_lib.sh]: In the case where `ADD_PKGS =""` wouldn't `INCLUDE_EXTRA` resolve to just  `"--include="`?
Shouldn't this condition be negated?

**Comment by amshinde** [comment]
@jcvenegas Can you take a look?

**Comment by jodh-intel** [comment]
Now that #144 is merged, can we close this one?

**Comment by grahamwhaley** [comment]
I concur, #144 looks to add Debian as well as Ubuntu. Closing this one @mcgrof  Feel free to test #144 and re-open/ammend this PR if you find something missing. Thanks everybody!

**Pull request closed without merging**